### PR TITLE
add recursive_version_locations option for searching revision files

### PR DESCRIPTION
**Pull request by ostr00000**
### Description

These changes implement recursive search for revision files. A new flag `recursive_version_locations` is added to config file `alembic.ini`.
Each `version_locations` path is separately searched recursively. Existing flag `sourceless` works as before this change.

As described in #760, the use case is to allow a user to manually reorganize his version location folder. It is possible to reorganize file structure at any time, without changing the configuration file. New revision files are created without any changes.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

Fixes: #760

### CODE REVIEW

**Comment by CaselIT** [comment]
Thanks for the PR,

We will review it soon!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 971156555abd2a76dd79253fd6b7523fce7ca3c3 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by zzzeek** [comment]
this is a great effort, thanks for the thoroughness.  im traveling this week and have limited time to look at things.   this would be for 1.10

File: alembic/script/base.py [hunk 1]
```diff
@@ -959,26 +966,40 @@ def _from_path(
...
-                paths.extend(
-                    os.path.join("__pycache__", pyc)
-                    for pyc in os.listdir(os.path.join(path, "__pycache__"))
-                    if pyc.split(".")[0] not in names
-                )
-            return paths
-        else:
-            return os.listdir(path)
+        paths = []
+        for root, dirs, files in os.walk(path):
```
Comment (sqla-tester) [inline alembic/script/base.py:970]: **mike bayer** (zzzeek) wrote:

it might be a good idea to indicate topdown=True here, even though this is default, to document / assert that the algo here relies upon the topdown walking behavior when it breaks out for non-recursive.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

File: tests/test_script_production.py [hunk 2]
```diff
@@ -1356,3 +1361,256 @@ def normpath(path):
...
+        rev_hash = self.rev[rev_index]
+        script = self.rev_to_script[rev_hash]
+        file_name = os.path.basename(script.path)
+        target_file = os.path.join(
+            _get_staging_directory(), version_path, destination_dir, file_name
+        )
+        target_file = os.path.realpath(target_file)
+        return target_file
+
+    def _checkTopologicalOrder(self, sd: ScriptDirectory):
```
Comment (sqla-tester) [inline tests/test_script_production.py:1400]: **mike bayer** (zzzeek) wrote:

nit: replace _lowerCamelCase with _underscore_and_dashes ; I guess the convention here is from python unittest but we aren't standardizing on that

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

File: tests/test_script_production.py [hunk 3]
```diff
@@ -1356,3 +1361,256 @@
...
+        }
+
+    def tearDown(self):
+        clear_staging_env()
+
+    def test_flat_structure(self):
+        assert len(self.rev) == 6
+
+    def test_flat_and_dir_structure(self):
+        self._move_revision_file(0, "dir_1")
```
Comment (sqla-tester) [inline tests/test_script_production.py:1453]: **mike bayer** (zzzeek) wrote:

i see this is part of the setup, but it's a little hard to follow that this is part of the setup which also builds on the three_rev_fixture / multi_heads_fixture.   also just generally we have some newer pytest-style conventions that are IMO easier to follow.


for clarity, can we change the fixture organization to be like this, using @testing.fixture (which is a facade over pytest.fixture):


    @testing.fixture
    def base_rev_fixture(self):
        rev = list(three_rev_fixture(self.cfg))
        rev.extend(multi_heads_fixture(self.cfg, *self.rev[0:3]))
        return rev

    @testing.fixture
    def flat_and_dir_fixture(self, base_rev_fixture):
        # ideally these wouldn't be attached to "self" in the test,
        # but I see _move_revision_file and _get_moved_path are relying
        # on these as instance variables for now, OK
        org_script_dir = self.org_script_dir = ScriptDirectory.from_config(self.cfg)
        rev_to_script = self.rev_to_script = {
                script.revision: script
                for script in org_script_dir.walk_revisions()
            }
        self._move_revision_file(0, "dir_1")
        self._move_revision_file(2, "dir_1")
        self._move_revision_file(4, "dir_2")
        self._move_revision_file(5, "dir_3")

        return org_script_dir

    def test_flat_and_dir_structure(self, flat_and_dir_fixture):
        sd = flat_and_dir_fixture

        # ... rest of test

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

this is great!   a couple of changes if you have time.    Otherwise I can take it over and push it through, let me know.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

File: alembic/script/base.py (hunk 1 above)
Comment (ostr00000) [reply alembic/script/base.py]: added explicit `topdown=True`

File: tests/test_script_production.py (hunk 2 above)
Comment (ostr00000) [reply tests/test_script_production.py:1400]: I use daily camelCase, so I must overlook this  - fixed.

File: tests/test_script_production.py (hunk 3 above)
Comment (ostr00000) [reply tests/test_script_production.py:1453]: I am not sure if I understood correctly this example. In the code, I extracted only `self._move_revision_file(0, "dir_1")` part into fixtures, because `base_rev_fixture, org_script_dir, rev_to_script` are common for tests in a given class.

I am aware that this part of code is somewhat tangled, so I tried to refactor it moving to common code to a new class `RevisionMover`, but in the end it turns out there was no readability improvement, so I stayed with current version (with fixtures).

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision ecb0da48b459abd3f5e95390ec7030a7e3fcbc6d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset ecb0da48b459abd3f5e95390ec7030a7e3fcbc6d added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

OK I've reworked the test suite further and also moved it into test_script_consumption.py, since this has to do with reading the files.    The tests are now under a single class and each test takes advantage of methods that write out / test a whole directory at once.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

**Comment by ostr00000** [comment]
This looks much better now! I also see that I missed one very important test (`test_ignore_for_non_recursive`). 
What is the current state? Are we waiting for a review?
I assume that failed tests in Gerrit is maybe some config problem and the failed test in github is not valid anymore (however, I have no idea why it failed).

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457 has been **merged**. Congratulations! :)
